internal/executor/executor.go

```diff
@@ -114,26 +114,25 @@ func (executor *Executor) RunBuild() {
 			}
 		}
 	}
-EXECUTION_LOOP:
+
+	// Even through we haven't run any commands yet, we bootstrap with the clean state
+	previousCommandSucceeded := true
+
 	for {
 		if currentStep.Name == executor.commandTo {
 			break
 		}
-		log.Printf("Executing %s...", currentStep.Name)
-		nextCommandName := executor.performStep(environment, currentStep)
-		if nextCommandName == "" {
-			log.Printf("%s command finished and instructed to exit!", currentStep.Name)
+
+		shouldRun := (currentStep.ExecutionBehaviour == api.Command_ON_SUCCESS && previousCommandSucceeded) ||
+			(currentStep.ExecutionBehaviour == api.Command_ON_FAILURE && !previousCommandSucceeded) ||
+			currentStep.ExecutionBehaviour == api.Command_ALWAYS
+		if !shouldRun {
 			break
 		}
+
+		log.Printf("Executing %s...", currentStep.Name)
```

Comment:
I don't see where the `currentStep` is changed.

Reply:
Fixed in https://github.com/cirruslabs/cirrus-ci-agent/pull/32/commits/baaa8762898f60f248c4476f5b86e6a9a0d1a5c1.